# Titanic Dataset — First Exploratory Data Analysis (EDA)
**Track:** Neurofive ML Track — Week 1
**Goal:** Set up the Python/Jupyter toolkit and "listen" to the Titanic dataset before doing any modeling.

Dataset source (direct CSV, no Kaggle login needed):
`https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv`
(This is a standard mirror of the Kaggle "Titanic - Machine Learning from Disaster" `train.csv`.)


## 1. Imports

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
print("pandas:", pd.__version__)
print("numpy:", np.__version__)


pandas: 3.0.2
numpy: 2.4.4


## 2. Load the dataset directly from GitHub

In [3]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
df.shape


(891, 12)

## 3. First look — `.head()`

In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 4. Structure and dtypes — `.info()`

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


## 5. Summary statistics — `.describe()`

In [5]:
df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
PassengerId,891.0,NaN,NaN,NaN,446.0,257.353842,1.0,223.5,446.0,668.5,891.0
Survived,891.0,NaN,NaN,NaN,0.383838,0.486592,0.0,0.0,0.0,1.0,1.0
Pclass,891.0,NaN,NaN,NaN,2.308642,0.836071,1.0,2.0,3.0,3.0,3.0
Name,891,891,"Braund, Mr. Owen Harris",1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Sex,891,2,male,577,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Age,714.0,NaN,NaN,NaN,29.699118,14.526497,0.42,20.125,28.0,38.0,80.0
SibSp,891.0,NaN,NaN,NaN,0.523008,1.102743,0.0,0.0,0.0,1.0,8.0
Parch,891.0,NaN,NaN,NaN,0.381594,0.806057,0.0,0.0,0.0,0.0,6.0
Ticket,891,681,347082,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Fare,891.0,NaN,NaN,NaN,32.204208,49.693429,0.0,7.9104,14.4542,31.0,512.3292


## 6. Missing values per column

In [6]:
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
missing_summary = pd.DataFrame({'missing_count': missing, 'missing_pct': missing_pct})
missing_summary[missing_summary['missing_count'] > 0].sort_values('missing_count', ascending=False)


,missing_count,missing_pct
Cabin,687,77.10
Age,177,19.87
Embarked,2,0.22


## 7. Categorical vs. numerical columns

In [7]:
numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

print("Numerical columns:", numerical_cols)
print("Categorical columns:", categorical_cols)


Numerical columns: ['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']
Categorical columns: ['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']


/tmp/ipykernel_557/2379709948.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns.tolist()


## 8. My First Data Story 

- The dataset has **891 rows and 12 columns**, one row per Titanic passenger.
- Three columns have missing values: **`Cabin`** is missing for ~77% of passengers, **`Age`** is missing for ~20%, and **`Embarked`** is missing for just 2 rows — `Cabin` is nearly unusable as-is, while `Age` is worth imputing carefully.
- **Numerical** columns: `PassengerId`, `Survived`, `Pclass`, `Age`, `SibSp`, `Parch`, `Fare` (7 columns). **Categorical** columns: `Name`, `Sex`, `Ticket`, `Cabin`, `Embarked` (5 columns) — note `Pclass` and `Survived` are numeric but really behave like categories.
- The target column `Survived` is binary (0/1), with survival rate around 38% — the dataset is imbalanced, which will matter for modeling later.
- `Fare` and `Age` are right-skewed (a few very high fares, e.g. up to \$512), so scaling/outlier handling will likely be needed before modeling.
- **Next step:** decide an imputation strategy for `Age`, drop or engineer a feature from `Cabin` (e.g. "has_cabin"), and one-hot encode `Sex` and `Embarked`.
